test/PaymentRelay.test.ts

```diff
@@ -26,94 +27,241 @@ contract("PaymentRelay", (accounts) => {
+            sender,
+            PAYMENT_UID,
+            receiver,
+            { from: anyUser }
           );
-          assert.fail("operatorSend() did not throw.");
+          assert.fail("execPayment() did not throw.");
         } catch (e: any) {
-          expect(e.message).to.includes("missing role");
+          expect(e.message).to.includes(
+            "Caller does not have permission to execute payment"
+          );
+        }
+      });
+
+      it("Should only allow contract operator to refund payment", async function () {
+        try {
+          await paymentRelayContract.refundPayment(sender, PAYMENT_UID, {
+            from: anyUser,
+          });
+          assert.fail("refundPayment() did not throw.");
+        } catch (e: any) {
+          expect(e.message).to.includes(
+            "Caller does not have permission to refund payment"
+          );
         }
+
+        try {
+          await paymentRelayContract.refundPayment(sender, PAYMENT_UID, {
+            from: sender,
+          });
+          assert.fail("refundPayment() did not throw.");
+        } catch (e: any) {
+          expect(e.message).to.includes(
+            "Caller does not have permission to refund payment"
+          );
+        }
+      });
+    });
+
+    describe("When payment is executed", function () {
+      before(async function () {
+        await paymentRelayContract.execPayment(sender, PAYMENT_UID, receiver, {
+          from: operator,
+        });
+      });
+
+      it("Should sent tokens", async function () {
+        expect((await champContract.balanceOf(receiver)).toString()).to.equals(
+          amount
+        );
+        expect(
+          (
+            await champContract.balanceOf(paymentRelayContract.address)
+          ).toString()
+        ).to.equals("0");
+      });
+
+      it("Should mark payment as executed", async function () {
+        expect(
+          await paymentRelayContract.isPaymentProcessed(PAYMENT_UID, sender)
+        ).to.be.true;
+
+        const { 2: paymentState } = await paymentRelayContract.getPayment(
+          PAYMENT_UID,
+          sender
+        );
+        expect(paymentState).to.be.equal(
+          await paymentRelayContract.PAYMENT_EXECUTED()
+        );
       });
 
       it("Should keep free other UID", async function () {
-        const ANY_UID = web3.utils.keccak256("ANY_UID");
+        expect(await paymentRelayContract.isPaymentReserved(ANY_UID, sender)).to
+          .be.false;
         expect(await paymentRelayContract.isPaymentProcessed(ANY_UID, sender))
           .to.be.false;
       });
 
       it("Should prevent UID collision between users", async function () {
+        expect(
+          await paymentRelayContract.isPaymentReserved(PAYMENT_UID, anyUser)
+        ).to.be.false;
         expect(
           await paymentRelayContract.isPaymentProcessed(PAYMENT_UID, anyUser)
         ).to.be.false;
       });
     });
+
+    describe("When payment is refunded", function () {
```

Comment:
P1, WDYT? 

Missing cases:

- should prevent to execute payment
- should prevent to refund payment again
- should prevent to reserve a payment with same UID

Reply:
Same comment and [this one](https://github.com/Unagi-Games/evm-vested/pull/4/files#r1153229127).